PEP 817: Wheel Variants: Beyond Platform Tags (#4740)
python/peps: peps/pep-0817.rst

              various targets. The first two bars are labeled "yum (2018.8)"
              and "generic (SSE2)", reach about 1.0 ns/day and are both
              marked as "SSE2". The next bar is labeled "ivybridge" ("AVX")
              and reaches almost 1.5 ns/day.  Two following bars are labeled
              "haswell" and "broadwell" (both "AVX2") and exceed 1.5 ns/day
              slightly.  The last two bars are labeled "skylake_avx512" and
              "cascadelake" (both "AVX512") and reach almost 2.0 ns/day.

        Performance of GROMACS 2020.1 built for different generations of
        CPUs.  Vertical axis shows performance expressed in ns/day, a
        GROMACS-specific measure of simulation speed (higher is better).

    “Compiling `GROMACS <https://www.gromacs.org/>`_ for architectures
    that can exploit the AVX-512 instructions supported by the Intel
    Cascade Lake microarchitecture gives an additional 18% performance
    improvement relative to using AVX2 instructions, with a speedup of
    about 70% compared to a generic GROMACS installation with only SSE2.”

    — `archspec: A library for detecting, labeling, and reasoning about

>     microarchitectures <https://tgamblin.github.io/pubs/archspec-canopie-hpc-2020.pdf>`__

Reviewer: Is this intended to be rendered as a block quote?

          provide the choice of PyTorch Build (stable or nightly),
          operating system (Linux, Mac, Windows), package (Pip,
          LibTorch, Source), language (Python, C++ / Java), and Compute
          Platform (CUDA 12.6, CUDA 12.8, CUDA 13.0, ROCM 6.4, CPU).
          Below these rows, the pip install command for the selected
          variant is provided, utilizing the --index-url parameter.

    The PyTorch install selector (https://pytorch.org/get-started/locally/,
    captured 22-Aug-2025)

PyTorch uses a combination of index URLs per accelerator type and local
version segments as accelerator tag (such as ``+cu130``, ``+rocm6.4`` or
``+cpu``) . Users need to first determine the correct index URL for their
system, and add an index specifically for PyTorch.

.. code:: bash

    pip install torch --index-url https://download.pytorch.org/whl/cu129

Tools need to implement special handling for the way PyTorch uses local
version segments. These requirements break the pattern that packages
are usually installed with. Problems with installing PyTorch
are a very common point of user confusion. To quantify this, on

> 2025-12-05, 552 out of 8136, or 6.8%, of issues on the `uv's issue

Reviewer: ```suggestion
2025-12-05, 552 out of 8136, or 6.8%, of issues on `uv's issue
```

``numpy-cuda`` package that users will be lead to believe it’s a CUDA
variant of NumPy.

`CuPy <https://cupy.dev/>`_ had to build
a total of 52 different packages - all with different names - which
clearly highlights the limit of such an approach. End users need to
carefully read the CuPy installation documentation to figure out
which package they need. Continuously having to create new PyPI
packages, request increasing limits and keep the build infrastructure
and documentation in sync, requires significant effort from software
maintainers.

.. code:: text

    cupy-cuda100 cupy-cuda101 cupy-cuda102 cupy-cuda110 cupy-cuda111 cupy-cuda112 cupy-cuda113 cupy-cuda114 cupy-cuda115
    cupy-cuda116 cupy-cuda117 cupy-cuda118 cupy-cuda119 cupy-cuda11x cupy-cuda120 cupy-cuda121 cupy-cuda122 cupy-cuda123
    cupy-cuda124 cupy-cuda125 cupy-cuda126 cupy-cuda127 cupy-cuda128 cupy-cuda129 cupy-cuda12x cupy-cuda13x cupy-cuda70
    cupy-cuda75 cupy-cuda80 cupy-cuda90 cupy-cuda91 cupy-cuda92 cupy-rocm-4-0 cupy-rocm-4-1 cupy-rocm-4-2 cupy-rocm-4-3
    cupy-rocm-4-4 cupy-rocm-4-5 cupy-rocm-5-0 cupy-rocm-5-1 cupy-rocm-5-2 cupy-rocm-5-3 cupy-rocm-5-4 cupy-rocm-5-5

>     cupy-rocm-5-6 cupy-rocm-5-7 cupy-rocm-5-8 cupy-rocm-5-9 cupy-rocm-6-0 cupy-rocm-6-1 cupy-rocm-6-2 cupy-rocm-6-3

Reviewer: If it's useful to keep all 52 names (shock factor!), please wrap to 70-ish characters, as literal blocks aren't reflowed on smaller screens.

GPU acceleration, or to provide custom package variants, such as builds
against different dependency ABIs. This is particularly challenging for
the scientific computing, artificial intelligence (AI), machine learning
(ML), and high-performance computing (HPC) communities.

This PEP proposes "Wheel Variants", an extension to the
:doc:`packaging:specifications/binary-distribution-format`. This
extension introduces a mechanism for package maintainers to declare
multiple build variants for the same package version, while allowing
installers to automatically select the most appropriate variant based on
system hardware and software characteristics. More specifically, it
proposes:

- An evolution of the wheel format called **Wheel Variant** that allows
  wheels to be distinguished by hardware or software attributes.

- A **variant provider plugin** interface that allows installers to
  dynamically detect platform attributes and select the most suitable
  wheel.

The goal is for a familiar ``[uv] pip install <package>`` to provide
the best user experience.

Motivation
==========

The Python packaging ecosystem has evolved to support increasingly

> diverse computing environments. The current software ecosystem often

Reviewer: "software ecosystem" here is software in general rather than just Python? Would be useful to disambiguate from "packaging ecosystem" in the previous sentence.

currently be included in the wheel tag. Having to build multiple wheels
targeting very different hardware configurations forces maintainers into
various distribution strategies that are suboptimal, and create friction
for users and authors of other software who wish to depend on the
package in question.

A few existing approaches are explored in the subsequent subsections.
These include maintaining separate package indexes for different
hardware configurations, bundling all potential variants into a single
wheel of considerable size, or using separate package names
(``mypackage-gpu``, ``mypackage-cpu``, etc.). Each of these approaches
has significant drawbacks.

According to the `2024 Python Developers Survey
<https://lp.jetbrains.com/python-developers-survey-2024/#purposes-for-using-python>`__,
a significant portion of respondents over the last years have been
successively using Python for scientific computing purposes, covering
such areas as Data analysis (steadily over 40% respondents), Machine
learning (grown to 40% in 2024) and Data engineering (around 30%).
Many of these use cases are directly impacted by suboptimal

> packaging.

Reviewer: I'd make this para the start of Motivation, it provides a useful contextualisation
```suggestion
The `2024 Python Developers Survey`__ shows that a significant proportion of
Python's users have scientific computing use-cases. This includes data analysis
(40% of respondents), machine learning (30%), and data engineering (30%).
Many of these use-cases are directly impacted by suboptimal Python packaging.

__ https://lp.jetbrains.com/python-developers-survey-2024/#purposes-for-using-python
```

> This issue is often crossing the boundaries of scientific computing.

Reviewer 2: It means that not only "scientific computing" packages (i.e. the groups mentioned below) are affected, but technically any package using extensions could be.

              (2018.8)" and "generic (SSE2)", reach about 1.0 ns/day and
              are both marked as "SSE2". The next bar is labeled
              "ivybridge" ("AVX") and reaches almost 1.5 ns/day. Two
              following bars are labeled "haswell" and "broadwell" (both
              "AVX2") and exceed 1.5 ns/day slightly. The last two bars
              are labeled "skylake_avx512" and "cascadelake" (both
              "AVX512") and reach almost 2.0 ns/day.

        Performance of GROMACS 2020.1 built for different generations of
        CPUs. Vertical axis shows performance expressed in ns/day, a
        GROMACS-specific measure of simulation speed (higher is better).

    Compiling `GROMACS <https://www.gromacs.org/>`_ for architectures
    that can exploit the AVX-512 instructions supported by the Intel
    Cascade Lake microarchitecture gives an additional 18% performance
    improvement relative to using AVX2 instructions, with a speedup of
    about 70% compared to a generic GROMACS installation with only SSE2.

    — `archspec: A library for detecting, labeling, and reasoning about
    microarchitectures

>     <https://tgamblin.github.io/pubs/archspec-canopie-hpc-2020.pdf>`__

Reviewer: Would suggest making this entire part it's own subsection, moving to the appendix, or deleting. 

I think the general argument that packaging can't target specific hardware beyond the most generic CPU architecture is already compelling, this text doesn't add much. That said, the general argument could be made more clearly and in a distilled form.

          provide the choice of PyTorch Build (stable or nightly),
          operating system (Linux, Mac, Windows), package (Pip,
          LibTorch, Source), language (Python, C++ / Java), and Compute
          Platform (CUDA 12.6, CUDA 12.8, CUDA 13.0, ROCM 6.4, CPU).
          Below these rows, the pip install command for the selected
          variant is provided, utilizing the --index-url parameter.

    The PyTorch install selector
    (https://pytorch.org/get-started/locally/, captured 22-Aug-2025)

PyTorch uses a combination of index URLs per accelerator type and local
version segments as accelerator tag (such as ``+cu130``, ``+rocm6.4`` or
``+cpu``) . Users need to first determine the correct index URL for
their system, and add an index specifically for PyTorch.

.. code:: bash

    pip install torch --index-url https://download.pytorch.org/whl/cu129

Tools need to implement special handling for the way PyTorch uses local
version segments. These requirements break the pattern that packages
are usually installed with. Problems with installing PyTorch
are a very common point of user confusion. To quantify this, on

> 2025-12-05, 552 out of 8136, or 6.8%, of issues on `uv's issue tracker

Reviewer: ```suggestion
2025-12-05, 552 out of 8136 (6.8%), of issues on `uv's issue tracker
```

<https://github.com/astral-sh/uv/issues/>`__ contained the term "torch".

Wheel variants remove this special casing and make GPU and TPU packages
work just as well as regular packages with native code. They also reduce
the burden on the maintainers to run separate infrastructure and to find
and use non-standard features such as local version segments present on

> an index.

Reviewer: Would delete this para. The _Motivation_ section should establish why the PEP/change is needed, rather than going into the solution. Regardless I don't think this text adds much / it applies to all the workaround cases equally

package names to approximate variants:

.. code:: bash

    pip install xgboost      # NVIDIA GPU variant
    pip install xgboost-cpu  # CPU-only variant

Maintainers of other software cannot express that they depend on either
of the available variants being selected. They need to
either depend on a specific variant, provide multiple alternative
dependency sets using extras, or even publish their own software using
multiple package names matching upstream variants.

Commonly, these packages install overlapping files. Since Python
packaging does not support expressing that two packages are mutually
exclusive, installers can install both of them to the same environment,
with the package installed second overwriting files from the one
installed first. This leads to runtime errors, and
the possibility of incidentally switching between variants depending on
the way package upgrades are ordered.

An additional limitation of this approach is that publishing a new
release synchronously across multiple package names is not currently
possible. :pep:`694` proposes adding such a mechanism for multiple
wheels within a single package, but extending it to multiple packages is

> not a goal.

Reviewer: This para feels less relevant, having a short delay between uploading packages is an annoyance rather than a usability issue on the scale of the rest of the PEP. Would delete for length.

**Induced Security Risk:** proliferation of suffixed variant packages
leads users to expect these suffixes in other packages, making name
squatting much easier. For example, one could create a malicious
``numpy-cuda`` package that users will be lead to believe it's a CUDA
variant of NumPy.

> `CuPy <https://cupy.dev/>`_ had to build

Reviewer: I assume building 52 packages is per release rather than just once in the project history
```suggestion
As at the time of writing, `CuPy <https://cupy.dev/>`__ has to build
```

Reviewer 2: No, it's a sum total of all packages ever built (until the time of writing). At any particular time, a different range of packages was built, so every package covers a different range of versions, and most of them are no longer updated.

a total of 52 different packages - all with different names - which
clearly highlights the limit of such an approach. End users need to
carefully read the CuPy installation documentation to figure out
which package they need. Continuously having to create new PyPI
packages, request increasing limits and keep the build infrastructure
and documentation in sync, requires significant effort from software

> maintainers.

Reviewer: ```suggestion
and documentation in sync, requires significant effort from maintainers.
```

    cupy-cuda92 cupy-cuda100 cupy-cuda101 cupy-cuda102 cupy-cuda110
    cupy-cuda111 cupy-cuda112 cupy-cuda113 cupy-cuda114 cupy-cuda115
    cupy-cuda116 cupy-cuda117 cupy-cuda118 cupy-cuda119 cupy-cuda11x
    cupy-cuda120 cupy-cuda121 cupy-cuda122 cupy-cuda123 cupy-cuda124
    cupy-cuda125 cupy-cuda126 cupy-cuda127 cupy-cuda128 cupy-cuda129
    cupy-cuda12x cupy-cuda13x
    cupy-rocm-4-0 cupy-rocm-4-1 cupy-rocm-4-2 cupy-rocm-4-3
    cupy-rocm-4-4 cupy-rocm-4-5 cupy-rocm-5-0 cupy-rocm-5-1
    cupy-rocm-5-2 cupy-rocm-5-3 cupy-rocm-5-4 cupy-rocm-5-5
    cupy-rocm-5-6 cupy-rocm-5-7 cupy-rocm-5-8 cupy-rocm-5-9
    cupy-rocm-6-0 cupy-rocm-6-1 cupy-rocm-6-2 cupy-rocm-6-3

Extra-Dependency as variants
''''''''''''''''''''''''''''

`JAX <https://docs.jax.dev/en/latest/installation.html>`__ uses a
plugin-based approach. The central ``jax`` package provides a number of
extras that can be used to install additional plugins,
e.g. ``jax[cuda12]`` or ``jax[tpu]``. This is far from ideal as
``pip install jax`` (with no extra) leads to a nonfunctional
installation, and consequently dependency chains, a fundamental expected
behavior in the Python ecosystem, are dysfunctional.

> JAX includes 12 extra selectors to cover all use cases - many of which

Reviewer: ```suggestion
JAX includes 12 extras to cover all use cases - many of which
```

overlap and could be misleading to users if they don't read the
documentation in detail.

It should be noted that most of these "extras" are technically mutually
exclusive, though it is currently impossible to correctly express this

> within the package metadata.

Reviewer: These two paras should be combined & hopefully shortened

prepared build environment that contains the ``torch`` package matching
the version that the software will run against, which requires building
without build isolation. On the project side, it requires hosting wheels
separately.

**Induced Security Risk:** Similar to regular source builds, this
model requires running arbitrary code at install time. The wheels
are downloaded entirely outside package manager's control, extending
the attack surface to two separate wheel download implementations and
preventing proper provenance tracking.

Ecosystem fragmentation
'''''''''''''''''''''''

The lack of standardized variant support has led to ecosystem
fragmentation:

**Inconsistent User Experience**: Each package uses different
installation methods, creating confusion and reducing discoverability.

**Development Tool Complications**: Installation tools, IDEs, and CI/CD
systems struggle to handle non-standard installation requirements.

**Error-proneness**: The established workarounds are often error-prone,
and in the past they have lead to issues such as downloading incorrect
artifacts.

Impact on scientific computing and AI/ML workflows

> --------------------------------------------------

Reviewer 3: Is there a requirement to keep these sections short, given that this is the non-normative part and the living spec is at packaging.python.org, where those sections won't be copied to?

I consider the quotes important for someone trying to evaluate the PEP. We're proposing a massive new addition, so we need to justify that it's worth it, and the testimonials from major Python projects and communities show how big the need is we're serving. If we're not sharing this in the PEP, we have to write this on DPO instead, where it's just way less accessible, because those questions will come up and the community work is a major part of this proposal.

> --------------------------------------------------

Reviewer 4: I'd agree with @konstin's comment. We thought about this pretty hard, and the in-depth motivation/rationale sections including quotes seem important, given:

1. The large amount of discussion leading up to this PEP over several years in many threads/places, and many including some regular packaging participants not recognizing the impact/important of this topic on the usability of Python as a whole for large parts of the end user population,
2. That this is a large proposal with above-average implications in the amount of work that various stakeholders (e.g., installer and build backend authors) need to do (asking for more work needs stronger motivation), and
3. That PEP 703 (free-threading), which is also a proposal with a huge scope and wide impact, set an excellent example with its usage of quotes.

- Maintains full backward compatibility with existing tools (i.e.,
  non-variant aware installers, tools, and indexes remain unaffected).
- Simplifies package maintenance by offering a unified and flexible
  solution to the challenge of managing multiple platform-specific
  package builds and distributions.
- Provides a seamless and predictable experience for users, with
  little-to-no user input required.
- Supports the full spectrum of modern computing hardware.
- Provides a future-proof and flexible system that can evolve with the
  ecosystem and future use cases.

Out-of-scope features
---------------------

This PEP tries to present the minimal scope required and leaves aspects
to tools to evolve. A non-exhaustive list:

- The format of a static file to select variants deterministically or
  include variants in a ``pylock.toml`` file,
- The list of variant providers that are vendored or re-implemented by
  installers,
- The specific opt-in mechanisms and UX for allowing an installer to run
  non-vendored variant providers,
- How to instruct build backends to emit variants through the :pep:`517`
  mechanism.

Prior art

> ---------

Reviewer: Move this entire section to an appendix or delete, I don't think it serves to significantly develop the text of the PEP

Reviewer 3: I consider learning from other ecosystems, their experiences and their mistakes a crucial part of any new design. It's important to consider if can adopt an existing, working solution from a different ecosystem with the same problems. The links in these sections were used as references when developing our own design. I'm not particular on placement, but this is not something we can delete.

API and ABI matching is primarily done through use of `slotting
<https://devmanual.gentoo.org/general-concepts/slotting/index.html>`__.
Slots are generally used to provide multiple versions or variants of
given package that can be installed alongside (e.g. different major GTK+
or LLVM versions, or GTK+3 and GTK4 builds of WebKitGTK), whereas
subslots are used to group versions within a slot, usually corresponding
to the library ABI version. Packages can then declare dependencies bound
to the slot and subslot used at build time. Again, separate binary
packages can be built against different dependency slots. When
installing a dependency version falling into a different slot or
subslot, the package manager may either replace the package needing that
dependency with a binary packages built against the new slot, or rebuild
it from source.

Normally, the use of slots assumes that upgrading to the newest version
possible is desirable. When more fine-grained control is desired, slots
are used in conjunction with USE flags. For example,
``llvm_slot_{major}`` flags are used to select a LLVM major version to
build against.

Rationale

> =========

Reviewer: I would move much of the prose from Specification to the Rationale. It took me a few read throughs of the Rationale section before concepts started to make sense. A couple of paragraphs or bullet points on how the design of the whole system fits together would be of great benefit at the start here

Reviewer 2: Isn't the goal though, that people could actually implement the specification from reading "specification" section alone? Given the complexity of the proposal, I think it makes sense to include a more explicit introduction to all the concepts there.

Variant properties system
-------------------------

Variant properties serve the purpose of expressing the characteristics
of the variant. Unlike platform compatibility tags, they are stored in
the variant metadata and therefore do not affect the wheel filename
length. They follow a hierarchical key-value design, with the key
further broken into a namespace and a feature name. Namespaces are used
to group features defined by a single provider, and to avoid conflicts
should multiple providers define a feature with the same name. This
permits independent governance and evolution of every namespace.

The keys are restricted to lowercase letters, digits and underscores.
Uppercase characters are disallowed to avoid different spellings of the
same name. The character set for values is more relaxed, to permit
values resembling versions.

Variant features can be declared as allowing multiple values to be
present within a single variant wheel. If that is the case, these values
are matched as a logical disjunction, i.e. only a single value needs to
be compatible with the system for the wheel to be considered supported.

> On the other hand, features are treated conjunctively, i.e. all of them

Reviewer: I'd just say OR / AND (`any()`/`all()`) instead of formal logic terminology

need to be compatible. This provides some flexibility in designating
variant compatibility while avoiding having to implement a complete
boolean logic.

Variant properties are serialized into a structured 3-tuple format
inspired by Trove Classifiers in :pep:`301`:

::

    {namespace} :: {feature_name} :: {feature_value}

Null variant

> ------------

Reviewer: This section could do with a better overview / motivation. I think 'null' as the name is also confusing, as (implicitly?) there are min requirements for the 'base case'/fallback wheel. I'm also not sure this section flows best here ordering-wise

a. Install-time providers that require executing code to determine
   the wheel compatibility, and therefore satisfy the variants needing
   that.

b. Ahead-of-Time (AoT) providers that entirely rely on static metadata
   for the purpose of wheel selection, and that can be used for the
   other class of variants to avoid the additional problems posed
   by install-time providers.

For user convenience and better consistency across different packages,
AoT providers can also feature plugin packages, but they are only used
at build time (where executing code from additional packages is already
a risk) to verify the variant properties specified by the maintainer
and automatically generate the static metadata.

Both kinds of provider plugins expose the same API. However, an AoT
provider must always consider all valid variant properties supported,
and it must always return the same ordered list of supported properties
irrespective of the user system. All AoT providers can be technically
used as install-time providers, but not the other way around.

Plugin stability and versioning
-------------------------------

Given that calling provider plugins may be necessary while installing

> variant wheels published in the past, it is important that provider

Reviewer: All (currently installable) wheels are published in the past! What is this sentence meant to mean?

> plugin behavior remain stable within their lifetime. Ideally, no

Reviewer: > it is important that provider plugin behavior remain stable within their lifetime

Grammar here doesn't make sense. Also, what behaviours? Stable how?

properties previously supported should ever be removed.

If a breaking change needs to be performed, it is recommended to either
introduce a new provider package for that, or add a new plugin API
endpoint to the existing package. In both cases, it may be necessary to
preserve the old endpoint in minimal maintenance mode, to ensure that
old wheels can still be installed. The old endpoint can trigger
deprecation warnings in the ``get_all_configs()`` hook that is used when
building packages.

An alternative approach is to use semantic versioning to cut off
breaking changes. However, this relies on package authors reliably using

> caps on dependencies, as otherwise old wheels will start using

Reviewer: Capping dependencies is strongly reccomended against, especially for libraries, right?

incompatible plugin versions. This is already a problem with Python
build backends used today.

When vendoring or reimplementing plugins, installers need to follow
their current behavior. In particular, they should recognize the
relevant provider versions numbers, and possibly fall back to installing
the external plugin when the package in question is incompatible with
the installer's implementation.

Example use cases

> -----------------

Reviewer: Would suggest moving this to an appendix for length

PyTorch CPU/GPU variants
''''''''''''''''''''''''

As of October 2025, `PyTorch
<https://pytorch.org/get-started/locally/>`__ publishes builds a total
of seven variants for every release: a CPU-only variant, three CUDA
variants with different minimal CUDA runtime versions and supported
GPUs, two ROCm variants and a Linux XPU variant.

This setup could be improved using GPU/XPU plugins that query the
installed runtime version and installed GPUs/XPUs to filter out the
wheels for which the runtime is unavailable, it is too old or the user's
GPU is not supported, and order the remaining variants by the runtime
version. The CPU-only version is published as a null variant that is

> always supported.

Reviewer 2: I don't really understand what you're referring to. We're not mentioning build infrastructure at all in this section. We're merely shortly stating what the problem is (paragraph 1), how wheel variants could be used (paragraph 2) and how that improves UX (paragraph 3).

chooses the wheel for the newest runtime supported. Otherwise, it falls
back to the CPU-only variant. In the corner case when multiple
accelerators are available and supported, PyTorch package maintainers
indicate which one takes preference by default.

Optimized CPU variants
''''''''''''''''''''''

Wheel variants can be used to provide variants requiring specific CPU
extensions, beyond what platform tags currently provide. They can be
particularly helpful when runtime dispatching is impractical, when the
package relies on prebuilt components that use instructions above the
baseline, when availability of instruction sets implies library ABI
changes, or simply to benefit from compiler optimizations such as
auto-vectorization applied across the code base.

For example, an x86-64 CPU plugin can detect the capabilities for the
installed CPU, mapping them onto the appropriate x86-64 architecture
level and a set of extended instruction sets. Variant wheels indicate
which level and/or instruction sets are required. The installer filters
out variants that do not meet the requirements and select the best
optimized variant. A non-variant wheel can be used to represent the

> architecture baseline, if supported.

Reviewer: Earlier the suggestion was to use the `null` variant. Is this different advice? If not, could it be clarified?

OpenMP implementations, whereas using a consistent implementation across
the stack can avoid degrading performance through spawning too many
threads.

BLAS / LAPACK variants do not require a plugin at install time, since
all variants built for a particular platform are compatible with it.
Therefore, an ahead-of-time provider (with ``install-time = false``)
that provides a predefined set of BLAS / LAPACK library names can be
used. When the package is installed, normally the default variant is
used, but the user can explicitly select another one.

Debug package variants
''''''''''''''''''''''

A package may wish to provide a special debug-enabled builds for
debugging or CI purposes, in addition to the regular release build. For
this purpose, an optional ahead-of-time provider can be used
(``install-time = false`` with ``optional = true``), defining a custom
property for the debug builds. Since the provider is disabled by
default, users normally install the non-variant wheel providing the
release build. However, they can easily obtain the debug build by
enabling the optional provider or selecting the variant explicitly.

> Package ABI matching

Reviewer 5: Super excited about this one by the way :D

executed with elevated privileges, e.g. when installing packages for
multi-user systems, while the installed packages are only used with
regular user privileges afterwards. Therefore, variant provider plugins
could introduce a Remote Code Execution vulnerability with elevated
privileges.

A similar issue already exists in the packaging ecosystem when packages
are installed from source distributions, whereas build backends
and other build dependencies are installed and executed. However,
various tools operating purely on wheels, as well as users using
tool-specific options to disable use of source distributions,
have been relying on the assumption that no code external to the system
will be executed while resolving dependencies, installing a wheel or
otherwise processing it. To uphold this assumption, the proposal
explicitly requires that provider plugin packages are never installed
without explicit user consent.

The `Overview`_ section of the specification provides further
suggestions that aim to improve both security and the user experience.
It is expected that a limited subset of popular provider plugins will
be either vendored by the installer, eliminating the use of packages

> external to the tool altogether, or pinned to specific versions,

Reviewer: The referenced section (Specification -> Overview) notes that pinning is to be avoided. Here, it is advised.

It is recommended that these tools vendor, reimplement or lock the most
commonly used plugins to specific specific wheels after verifying
their trustworthiness, to enable the ability to securely install variant
wheels out-of-the-box. To reduce the maintenance costs, repositories of
such vetted plugins could be maintained collaboratively and shared
between different tools.

For plugins and their dependencies not in such a pre-approved list,
a trust-on-first-use mechanism for every version is recommended. In
interactive sessions, the tool can explicitly ask the user for approval.
In non-interactive sessions, the approval can be given using
command-line interface options. It is important that the user is
informed of the risk before giving such an approval.

For a consistent experience between tools, variant wheels should be
supported by default. Tools may provide an option to only use
non-variant wheels. For scenarios requiring more control, providers can
be marked as optional and then must be explicitly enabled by the user
for the wheels to become available for installation.

Summary of changes
-------------------

> The Wheel Variant PEP introduces four key components:

Reviewer: Per previous comment, this is the sort of thing that could go into the introduction of _Rationale_

      i. ``*.dist-info/variant.json``: Individual wheel variant
         metadata.
      ii. ``*-variants.json``: Variant metadata file aggregated on the
          package index.

3. `Provider plugin API`_: Plugin interface to allow detection of
   system capabilities and validate variant compatibility.

4. `Variant environment markers`_: New environment markers to declare
   dependencies that are applicable to a subset of variants only.

Extended wheel filename
-----------------------

This PEP changes the wheel filename template originally defined by
:pep:`427` to:

.. code:: text

    {distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-{platform tag}(-{variant label})?.whl

Wheels using extensions introduced by this PEP must feature the variant
label component. The label must adhere to the following rules:

- Lower case only (to prevent issues with case-sensitive
  vs. case-insensitive filesystems)
- Between 1-16 characters
- Using only ``0-9``, ``a-z``, ``.`` or ``_`` ASCII characters

This is equivalent to the following regular expression:
``^[0-9a-z._]{1,16}$``.

Every label must uniquely correspond to a specific set of variant

> properties, same for all wheels using the same label within a single

Reviewer: ```suggestion
properties, which must be the same for all wheels using the same label within a single
```

All three variants metadata files share a common JSON-compatible
structure:

.. code:: text

    (root)
    |
    +- providers
    |  +- {namespace}
    |     +- enable-if     : str | None = None
    |     +- install-time  : bool       = True
    |     +- optional      : bool       = False
    |     +- plugin-api    : str | None = None
    |     +- requires      : list[str]  = []
    |
    +- default-priorities
    |  +- namespace        : list[str]
    |  +- feature
    |     +- {namespace}   : list[str]  = []
    |  +- property
    |     +- {namespace}
    |        +- {feature}  : list[str]  = []
    |
    +- static-properties
    |  +- {namespace}
    |     +- {feature}     : list[str]  = []
    |
    +- variants
      +- {variant_label}
         +- {namespace}
            +- {feature}   : list[str]  = []

The top-level object is a dictionary rooted at a specific point in the
containing file. Its individual keys are sub-dictionaries that are
described in the subsequent sections, along with the requirements for
their presence. The tools must ignore unknown keys in the dictionaries
to permit future backwards compatible updates to the PEP. However, users

> should not introduce custom keys to avoid potential future conflicts.

Reviewer: ```suggestion
MUST NOT use unsupported keys to avoid potential future conflicts.
```

dictionaries with provider information. It specifies how to install and
use variant providers. A provider information dictionary must be
declared in ``pyproject.toml`` for every variant namespace supported by
the package. It must be copied to ``variant.json`` as-is, including
the data for providers that are not used in the particular wheel.

A provider information dictionary may contain the following keys:

- ``enable-if: str``: An :ref:`environment marker
  <dependency-specifiers-environment-markers>` defining when the plugin
  should be used. If the environment marker does not match the running
  environment, the provider will be disabled and the variants using its
  properties will be deemed incompatible. If not provided, the plugin
  will be used in all environments.

- ``install-time: bool``: Whether this is an install-time provider.
  Defaults to ``true``. ``false`` means that it is an AoT provider
  instead.

- ``optional: bool``: Whether the provider is optional. Defaults
  to ``false``. If it is ``true``, the provider is
  considered optional and should not be used unless the user opts in to
  it, effectively rendering the variants using its properties

>   incompatible. If it is ``false``, the provider is considered

Reviewer: I didn't understand this

PEP: 817
Title: Wheel Variants: Beyond Platform Tags
Author: Jonathan Dekhtiar <jonathan@dekhtiar.com>,
        Michał Górny <mgorny@quansight.com>,
        Konstantin Schütze <konstin@mailbox.org>,
        Ralf Gommers <ralf.gommers@gmail.com>,
        Andrey Talman <atalman@meta.com>,
        Charlie Marsh <charlie@astral.sh>,
        Michael Sarahan <msarahan@gmail.com>,
        Eli Uriegas <eliuriegas@meta.com>,
        Barry Warsaw <barry@python.org>,
        Donald Stufft <donald@stufft.io>,
        Andy R. Terrel <andy.terrel@gmail.com>
Discussions-To: Pending
Status: Draft
Type: Standards Track
Topic: Packaging
Created: 10-Dec-2025
Post-History:

Abstract
========

The Python wheel packaging format uses
:doc:`packaging:specifications/platform-compatibility-tags` to specify
a given wheel's supported environments. These tags are unable to express
features of modern hardware configurations, such as the availability of
GPU acceleration, or to provide custom package variants, such as builds
against different dependency ABIs. This is particularly challenging for
the scientific computing, artificial intelligence (AI), machine learning

> (ML), and high-performance computing (HPC) communities.

Reviewer 6: ```suggestion
Python's existing wheel packaging format uses
:doc:`packaging:specifications/platform-compatibility-tags` to specify
a given wheel's supported environments. These tags are unable to express
modern hardware configurations and their features, such as the availability of
GPU acceleration. The tags fail to provide custom package variants, such as builds
against different dependency ABIs. These inabilities are particularly challenging for
scientific computing, artificial intelligence (AI), machine learning
(ML), and high-performance computing (HPC) communities.
```

   or newer system.

While these tags effectively handle traditional compatibility
dimensions, they cannot express modern requirements:

* **GPU Accelerated Frameworks:** A wheel filename like
``torch-2.9.0-cp313-cp313-manylinux_2_28_x86_64.whl`` provides no
indication whether it supports NVIDIA CUDA, AMD ROCm,
Intel XPU, or is CPU-only. Users cannot determine compatibility
with their GPU hardware or drivers.

* **CPU Instruction Sets:** A wheel filename like
``numpy-2.3.2-cp313-cp313-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl``
provides no indication whether it uses optimized instructions
for modern processors with AVX512, SHA-NI, and other specialized
instruction sets. Packages cannot indicate whether they
require specific CPU baseline. This forces maintainers to either use
inconvenient runtime dispatching, or rely on the lowest common
denominator, leaving performance improvements on the table.

* **Runtime Dependencies:** Scientific computing packages often depend on
specific BLAS implementations (OpenBLAS vs Intel MKL), MPI providers
(OpenMPI vs MPICH), or other system libraries that affect both
functionality and performance. The current wheel format is not able to

> encode that dependency.

Reviewer 7: Fix indent to fix build:
```suggestion
* **GPU Accelerated Frameworks:** A wheel filename like
  ``torch-2.9.0-cp313-cp313-manylinux_2_28_x86_64.whl`` provides no
  indication whether it supports NVIDIA CUDA, AMD ROCm,
  Intel XPU, or is CPU-only. Users cannot determine compatibility
  with their GPU hardware or drivers.

* **CPU Instruction Sets:** A wheel filename like
  ``numpy-2.3.2-cp313-cp313-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl``
  provides no indication whether it uses optimized instructions
  for modern processors with AVX512, SHA-NI, and other specialized
  instruction sets. Packages cannot indicate whether they
  require specific CPU baseline. This forces maintainers to either use
  inconvenient runtime dispatching, or rely on the lowest common
  denominator, leaving performance improvements on the table.

* **Runtime Dependencies:** Scientific computing packages often depend on
  specific BLAS implementations (OpenBLAS vs Intel MKL), MPI providers
  (OpenMPI vs MPICH), or other system libraries that affect both
  functionality and performance. The current wheel format is not able to
  encode that dependency.
```

    Wheel variants are a clear step in the right direction in this
    regard.

    — Michael Hudgins, JAX_ Developer Infrastructure Lead

They affect everyone from package authors to end users of all skill
levels, including students, scientists and engineers:

    Accessing compute to run models and process large datasets has been
    a pain point in scientific computing for over a decade. Today,
    researchers and data scientists still spend hours to days installing
    core tools like PyTorch before they can begin their work. This
    complexity is a significant barrier to entry for users who want to
    use Python in their daily work. The WheelNext Wheel Variants
    proposal offers a pathway to address persistent installation and
    compute-access problems within the broader packaging ecosystem
    without creating another, new and separate solution. Let's focus on
    the big picture of enhancing user experience - it will make a real
    difference.

    — Leah Wasser, Executive Director and Founder of `pyOpenSci
    <https://www.pyopensci.org/>`_

Heterogeneous computing environments
''''''''''''''''''''''''''''''''''''

> Research institutions and cloud providers often manage heterogeneous

Reviewer 6: ```suggestion
Research institutions and cloud providers manage heterogeneous
```

      runtimes (or none at all). So far, XGBoost has been vendoring a
      copy of OpenMP runtime, but this is increasingly untenable. Users
      get undefined behavior such as crashes or hangs when multiple
      incompatible versions of OpenMP runtimes are present in the
      system. (This problem was particularly bad on MacOS, so much so
      that the MacOS wheel for XGBoost no longer bundles OpenMP.)

    — Philip Hyunsu Cho, a lead maintainer of XGBoost_

The potential for improvement can be summarized as:

    This PEP is a significant step forward in improving the deployment
    challenges of the Python ecosystem in the face of increasingly
    complex and varied hardware configurations. By enabling multiple
    deployment targets for the same libraries in a standard way, it will
    consolidate and simplify many awkward and time-consuming
    work-arounds developers have been pursuing to support the rapidly
    growing AI/ML and scientific computing worlds.

    — Travis Oliphant, the author of NumPy_ and SciPy_ and Chief AI
    Architect at OpenTeams

Motivation summary
------------------

As highlighted in the previous section, the current Python packaging

> system cannot adequately serve the needs of modern heterogeneous

Reviewer 6: ```suggestion
system does not adequately serve the needs of modern heterogeneous
```
